### PR TITLE
Add lock free hash map.

### PR DESCRIPTION
**Pull request by toregge**
@baldersheim : please review

### CODE REVIEW

File: vespalib/src/vespa/vespalib/datastore/simple_hash_map.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include <vespa/vespalib/util/generationholder.h>
+#include <functional>
+
+namespace vespalib::datastore {
+
+class FixedSizeHashMap;
+class EntryComparator;
+
+/*
+ * Hash map over keys in data store.
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/simple_hash_map.h:16]: Please add more details on the implementation, how it relates to a data store, what is the key and value, and how to use it (e.g. how it relates to writer / readers and generation handling).

Also consider if the class name is descriptive enough.

File: vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+#include <deque>
+#include <functional>
+
+namespace vespalib { class GenerationHolder; }
+namespace vespalib::datastore {
+
+class EntryComparator;
+
+/*
+ * Fixed sized hash map over keys in data store.
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h:20]: Please add more details on the implementation, how it relates to a data store, what is the key and value, and how to use it (e.g. how it relates to writer / readers and generation handling).

File: vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        Node(Node &&rhs); // Must be defined, but must never be used.
+        void on_free();
+        std::atomic<uint32_t>& get_next() noexcept { return _next; }
+        const std::atomic<uint32_t>& get_next() const noexcept { return _next; }
+        KvType& get_kv() noexcept { return _kv; }
+        const KvType& get_kv() const noexcept { return _kv; }
+    };
+
+    Array<ChainHead>  _chain_heads;
```
Comment (baldersheim) [inline vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h:80]: I wonder if it is better to use std::vector<T, vespalib::allocator_large<T>> instead of Array. It should give the same advantages on allocation, and supports debugging better and is also more standard.
Ideally we should have something even simpler that does not require move or copy constructors.

**Review by baldersheim — APPROVED** [review]

File: vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                _nodes[prev_node_idx].get_next().store(next, std::memory_order_release);
+            } else {
+                chain_head.set(next);
+            }
+            --_count;
+            ++_hold_count;
+            _hold_1_list.push_back(node_idx);
+            return &_nodes[node_idx].get_kv();
+        }
+        node_idx = next;
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.cpp:150]: I think _prev_node_idx_ should be updated here as well. Otherwise the removed element cannot be removed from the node link chain as L140 never will trigger. I also think we should write separate unit tests for FixedSizeHashMap, such that all edge cases and intricacies are tested explicitly.

File: vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        // Writer thread
+        uint32_t load_relaxed() const noexcept { return _node_idx.load(std::memory_order_relaxed); }
+        void set(uint32_t node_idx) { _node_idx.store(node_idx, std::memory_order_release); }
+
+        // Reader thread
+        uint32_t load_acquire() const noexcept { return _node_idx.load(std::memory_order_acquire); }
+    };
+    class Node {
+        KvType _kv;
+        std::atomic<uint32_t> _next;
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h:61]: Consider rename to __next_node_idx_ or _next_idx_ for clarity.

File: vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    for (uint32_t node_idx : _hold_1_list) {
+        hold_2_list.push_back(std::make_pair(generation, node_idx));
+    }
+    _hold_1_list.clear();
+
+}
+
+
+void
+FixedSizeHashMap::trim_hold_lists_slow(generation_t usedGen)
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.cpp:110]: Consider rename usedGen -> first_used.

File: vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    KvType* remove(const EntryComparator& comp, EntryRef key_ref);
+    const KvType* find(const EntryComparator& comp, EntryRef key_ref) const;
+
+    void transfer_hold_lists(generation_t generation) {
+        if (!_hold_1_list.empty()) {
+            transfer_hold_lists_slow(generation);
+        }
+    }
+
+    void trim_hold_lists(generation_t usedGen) {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h:109]: Consider rename usedGen -> first_used.

File: vespalib/src/vespa/vespalib/datastore/simple_hash_map.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    void alloc_stripe(size_t stripe);
+    void hold_stripe(std::unique_ptr<const FixedSizeHashMap> map);
+public:
+    SimpleHashMap(std::unique_ptr<const EntryComparator> comp);
+    ~SimpleHashMap();
+    KvType& add(const EntryComparator& comp, std::function<EntryRef(void)> &insert_entry);
+    KvType* remove(const EntryComparator& comp, EntryRef key_ref);
+    const KvType* find(const EntryComparator& comp, EntryRef key_ref) const;
+    void transfer_hold_lists(generation_t generation);
+    void trim_hold_lists(generation_t used_gen);
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/simple_hash_map.h:53]: Consider rename used_gen -> first_used.

File: vespalib/src/tests/datastore/simple_hash_map/simple_hash_map_test.cpp [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+
+TEST_F(DataStoreSimpleHashTest, smoke_test)
+{
+    EXPECT_EQ(0, _hash_map.size());
+    insert(1);
+    EXPECT_EQ(1, _hash_map.size());
+    remove(2);
+    EXPECT_EQ(1, _hash_map.size());
```
Comment (geirst) [inline vespalib/src/tests/datastore/simple_hash_map/simple_hash_map_test.cpp:183]: Why is this not 0 as _count is subtracted at L145 in fixed_size_hash_map.cpp?

**Review by geirst — COMMENTED** [review]
Overall looks good. @toregge some comments to look at.
